Optimize pattern search in TileSet editor with HashMap lookup

killerdevildog opened the pull request:
Replace O(n) linear search with O(1) HashMap lookup for pattern preview updates in both TileMapLayerEditor and TileSetEditor.

- Add pattern_to_index_map HashMap to both editor classes
- Update _update_patterns_list() to populate the HashMap
- Optimize _pattern_preview_done() to use HashMap lookup
- Improves performance when working with many tile patterns

Resolves TODO comments about pattern search optimization.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

killerdevildog committed:
Optimize pattern search in TileSet editor with HashMap lookup

Replace O(n) linear search with O(1) HashMap lookup for pattern preview
updates in both TileMapLayerEditor and TileSetEditor.

- Add pattern_to_index_map HashMap to both editor classes
- Update _update_patterns_list() to populate the HashMap
- Optimize _pattern_preview_done() to use HashMap lookup
- Improves performance when working with many tile patterns

Resolves TODO comments about pattern search optimization.

killerdevildog committed:
Optimize TileMapPattern preview generation with caching and batch processing

- Add HashMap cache to store generated pattern previews for instant reuse
- Implement batch processing for pattern generation (process up to 4 patterns at once)
- Add cache hit check in queue_pattern_preview() for immediate texture return
- Add clear_pattern_preview_cache() method for cache management
- Significantly improves performance when loading patterns repeatedly
- Reduces SubViewport creation overhead by batching pattern processing

Performance improvement: Pattern previews now load instantly on cache hits,
eliminating expensive SubViewport rendering for previously seen patterns.

NOTE: benchmark_test_scenes/ directory included for maintainer reference only.
This directory contains before/after test scenes and benchmark video showing
the performance improvement. Please remove this directory before merge.

killerdevildog committed:
Optimize TileSet pattern loading with lazy pattern preview generation

This commit implements lazy loading for tile pattern previews to dramatically
improve startup performance when opening projects with many tile patterns.

Key optimizations implemented:

1. Lazy Pattern Preview Loading:
   - Remove immediate pattern preview generation during startup in _update_patterns_list()
   - Add placeholder icons (TileSet icon) for patterns until real previews are needed
   - Only generate previews when patterns become visible or when Patterns tab is accessed

2. Viewport-Based Rendering:
   - Add _schedule_pattern_preview_updates() method for intelligent preview scheduling
   - Check pattern visibility using viewport intersection before generating previews
   - Use call_deferred to ensure proper timing when switching to Patterns tab

3. Optimized Cache Integration:
   - Leverage existing pattern preview cache with getptr() optimization from previous commit
   - Maintain efficient HashMap-based pattern lookup (pattern_to_index_map)
   - Cache hit detection prevents redundant SubViewport creation
   - Eliminated double hash lookup (has() + []) using single getptr() call for better performance

4. Event-Driven Updates:
   - Connect visibility_changed and item_list_changed signals for responsive loading
   - Trigger lazy loading when Patterns tab becomes visible
   - Ensure smooth user experience with on-demand preview generation

Performance Impact:
- Eliminates expensive SubViewport creation during project startup
- Reduces initial loading time from several seconds to near-instant
- Maintains responsive UI while patterns load in background
- Optimized hash table operations reduce cache lookup overhead
- New benchmark video demonstrates significant startup improvement

Files Modified:
- tile_set_editor.h: Added _schedule_pattern_preview_updates() declaration
- tile_set_editor.cpp: Implemented lazy loading logic and viewport detection
- tiles_editor_plugin.cpp: Maintained cache optimization with getptr() improvement

Combined with the previous pattern preview caching optimization that eliminated
double hash lookups, this lazy loading approach provides comprehensive performance
improvements for TileSet pattern handling.

The benchmark video shows the new performance characteristics compared to
previous implementations, demonstrating the effectiveness of lazy loading
for tile pattern previews.

killerdevildog commented on editor/scene/2d/tiles/tile_map_layer_editor.cpp:
```diff
@@ -351,11 +351,11 @@ void TileMapLayerEditorTilesPlugin::_patterns_item_list_gui_input(const Ref<Inpu
-			patterns_item_list->set_item_icon(i, p_texture);
-			break;
+	// Optimized: Use HashMap for O(1) lookup instead of O(n) linear search
+	if (pattern_to_index_map.has(p_pattern)) {
```
I noticed that, I made the necessary change and a little more. please see my most recent commit.

killerdevildog replied:
please see most recent commit.

KoBeWi replied:
In File Changed tab I can see the most recent state of the files. This comment is still not addressed, you applied it somewhere else.

killerdevildog commented on benchmark_test_scenes/Benchmark_3rd commit.webm:
just the videos?

KoBeWi replied:
And zips.

killerdevildog replied:
Someone from the new-contributors told me to attach them to the pull request. ill update it.

KoBeWi replied:
You can attach a file by editing a post and drag and dropping the file.

killerdevildog committed:
Remove benchmark_test_scenes folder

- Remove all benchmark videos and test files
- Clean up repository by removing large video files
- Keep only the core optimization code changes

killerdevildog commented:
[Screencast from 2025-07-16 14-39-50.webm](https://github.com/user-attachments/assets/6ed601f2-45c1-49a7-8a94-a806e1eb34b1)


This video shows the passed behavior and fix.

killerdevildog committed:
Apply clang-format to TileMap editor files

- Fix code formatting in tile_set_editor.cpp
- Fix code formatting in tiles_editor_plugin.cpp and tiles_editor_plugin.h
- Ensure compliance with Godot coding standards

groud commented on editor/scene/2d/tiles/tiles_editor_plugin.h:
```diff
@@ -86,6 +93,7 @@ class TilesEditorUtils : public Object {
 
 	// Pattern preview API.
 	void queue_pattern_preview(Ref<TileSet> p_tile_set, Ref<TileMapPattern> p_pattern, Callable p_callback);
+	void clear_pattern_preview_cache();
```
This seems never called.

groud commented on editor/scene/2d/tiles/tiles_editor_plugin.cpp:
```diff
@@ -145,13 +162,34 @@ void TilesEditorUtils::_thread() {
 	ERR_FAIL_COND(p_pattern.is_null());
+
+	// Generate cache key based on pattern content and tileset
+	String cache_key = String::num_int64(p_tile_set->get_instance_id()) + "_" + String::num_int64(p_pattern->get_instance_id());
```
I am not sure this it enough. The problem is that, if the content of the tileset is modifed, then all stored preview will simply be outdated. It likely needs to be fully cleared if the TileSet is modified.
